Add a standard format for archiving build results.

nuclearsandwich opened the pull request:
The first commit adds documentation describing a proposed build-results-archive file.

Further commits will add to this python module API methods for building the described archive.

I'm opening this as a draft but I am looking for feedback on the proposed format.

nuclearsandwich committed:
Add documentation describing a proposed build-results-archive file.

Further commits will add to this python module API methods for building
the described archive.

mjeronimo commented on docs/build-results-archive.md:
```diff
@@ -0,0 +1,52 @@
+
+## build/**/*.sarif
+
+The build/ subdirectory _must_ contain the unaltered sarif content of the workspace build directory after the end of the latest colcon test run.
```
Should we also state that the files should be read-only, just to reinforce the idea that they are immutable?

nuclearsandwich replied:
I think that the archive itself should be entirely read-only and immutable. It's worth noting that in the text but whether the files themselves are "read-only" or not will be implementation dependent. (Ours will probably try to make them as such)

nuclearsandwich replied:
Noted in https://github.com/space-ros/process_sarif/pull/6/commits/b4440873d32415b983db5dbf23889cdfe06c9108 along with a suggestion that applications using the data use their own formats rather than alter these files.

mjeronimo commented on docs/build-results-archive.md:
```diff
@@ -0,0 +1,52 @@
+
+This file is identical in behavior to `colcon-build-cmd` but should describe the test command under the `latest_test` symlink.
+
+## processed-sarif-cmd
```
Once we have a "colcon sarif" command (whatever we decide the specific verb name should be), this could become "colcon-sarif-cmd", and would then be unform with "colcon-build-cmd", and "colcon-test-cmd".

nuclearsandwich replied:
It could although I'm not entirely convinced that it should be a colcon command or that even if it is a colcon command that it needs to be and therefore semantically I would argue against putting colcon in the filename. In fact I'd probably sooner argue that we should strip it from the build and test command filenames and call them `build-cmd` and `test-cmd` except that there really isn't a substitute for colcon in those cases and I think it's worth separating that definition from the build and test commands for any one package.

If we want to bikeshed a little perhaps `workspace-{build,test}-cmd`?

Ronoman commented on docs/build-results-archive.md:
```diff
@@ -0,0 +1,52 @@
+# Build results archive.
```
Should we specify where this results archive is written to?

mayman99 replied:
It is specified in the next section :+1:

nuclearsandwich replied:
@Ronoman as mentioned these lines deal with the path and filename conventions: https://github.com/space-ros/process_sarif/pull/6/files#diff-c55023a2efb41f70f1da8ee83deee86689f8bc5e843e745e305ccd8c750c5defR11-R15

Ronoman commented on docs/build-results-archive.md:
```diff
@@ -0,0 +1,52 @@
+## colcon-build-cmd
+
+This file _must_ contain the full colcon build command described in the colcon logs under the `latest_build` symlink.
+The file, when run as a /bin/sh program or prefixed with the `exec` shell command in a workspace root should invoke colcon in the local environment using the same arguments as the original build.
```
Nitpick: `should`, or `shall`?

Ronoman replied:
Won't repeat the review comment, but there are more instances of `should` in the doc to check. Also either all {should, shall, must} instances should be wrapped with `_`, or none of them (though happy to be overridden on this one).

nuclearsandwich replied:
Good catch on the should vs shall/must, I fixed that in https://github.com/space-ros/process_sarif/pull/6/commits/f3144142872728a02d0d7281cd6ae42dcaf9a5d0 along with emphasizing RFC words consistently.

Ronoman commented on docs/build-results-archive.md:
```diff
@@ -0,0 +1,52 @@
+If it is present, the processed.sarif file _must_ also be present.
+If present it should be identical to the colcon-build-cmd file in format but describe instead the command used to generate the processed.sarif file.
+
+## processed.sarif
```
Is this something we want to have? Merging Sarif files is non-trivial, but we could implement it.

Pros:
- Single file to open for the Dashboard to process and produce a view for
- Single place for developers to look for output if they want a manual view

Cons:
- A bit tricky to implement
    - What to do with multiple tools?
    - Need to merge Artifact lists and update Result indices accordingly
    - Need to ensure SARIF metadata is set properly (high level `properties` field)

It may be as simple as appending multiple objects to the `runs` list, but currently `process_sarif` assumes a single run per SARIF file.

mayman99 replied:
+1 for not merging the files

nuclearsandwich replied:
Okay it sounds like from out-loud discussion that keeping the same file structure (one output per tool per package) in the processed sarif is the way to go. This means that rather than have one processed.sarif file we'll have a directory mirroring the build directory but with de-duplicated sarif files.

nuclearsandwich replied:
https://github.com/space-ros/process_sarif/pull/6/commits/2d3b4fa30c36f6680f170652e8e1f198868ed6aa updates the doc to specify a processed directory rather than a single file.

nuclearsandwich committed:
Document that this file is expected to be read-only.

nuclearsandwich committed:
Emphasize RFC 2119 standard text.

nuclearsandwich committed:
Matches the directory structure of the build/ directory for processed SARIF.

Consolidating SARIF into a single file while previously discussed is not
a straightforward process and so maintaining the sarif file structure
from the original build, but still processing file contents, is the
preferred path.

nuclearsandwich committed:
Add path property to SarifFile.

nuclearsandwich committed:
Add console_script for generating a basic build_results_archive.

This is the first full spike of a build_results_archive generator.
Its current limitations are:

* It must implicitly be run from the workspace root.
* It takes no arguments to change the expected paths or names from their
  defaults.
* It's not cleaning up the temporary files that it creates to archive.
* It isn't timestamping the build_results_archive or creatng symlinks to
  a latest_build_results_archive.tar
* No compression support yet although bzip2 should be Fine:tm:.
* Internal API is not well factored.

nuclearsandwich committed:
Drop resources directory which no longer has contents.

nuclearsandwich committed:
Add console_script and use lexicographical sort order.

Commit:
Add repos exact into archive

Commit:
Swapped to get_sarif_in_build, added vcs exact to archive

Commit:
Remove todo

Ronoman commented:
Added vcs exact metadata, and swapped out the globbing list comprehension for a sarif_helpers function which handles malformed SARIF better.

nuclearsandwich commented on process_sarif/build_results_archive.py:
```diff
@@ -42,7 +43,15 @@ def main():
+        
         # TODO(Steven!) Add vcs-export-exact.repos
+        # Assume we are at the workspace root, since log/build_results_archives is relative to ws_root.
+        repos = subprocess.call(["vcs", "export", "--exact", "src"])
```
At the very least, it's probably worth making this a `check_call()` so that the entire process fails if the vcs-export does. I had been thinking to use vcs as a [runnable module](https://peps.python.org/pep-0338/) rather than shelling out directly but I'm not sure that it's ultimately that important

nuclearsandwich committed:
Add specified build and test command files.

nuclearsandwich committed:
Perform all cleanup at the end of the function.

At some point it may be worth it to try..finally this in case of other
errors.

nuclearsandwich committed:
Remove dead commented code.

nuclearsandwich committed:
Add the processed-sarif-cmd file to the archive.

nuclearsandwich committed:
Use standard timestamped filename.

This is inspired by the filenames used by colcon's existing log features
but is fully ISO8601 compliant and indicates that the time is in UTC.

nuclearsandwich committed:
Add sarif and processed sarif files relative to the current directory.

Otherwise the archive is not properly formatted.

nuclearsandwich committed:
bzip2 compress the archives.

nuclearsandwich committed:
This todo is resolved.

Since a colcon workspace has no *explicit* indicator there is no way I
know of to pre-emptively warn or or find the "workspace root" directory.
So it is now required that this command be run with the workspace root
as the current directory. The build and test command extraction
functions will fail if the current directory lacks a log/latest_build or
log/latest_test directory respectively which will enforce this
requirement.

nuclearsandwich committed:
Save the archive filename and path for later use.

nuclearsandwich committed:
Create a local symlink for the latest build result archive.

nuclearsandwich committed:
Add an additional title hierarchy for archive contents.

nuclearsandwich committed:
Add filename conventions and symlink behavior as optional.

nuclearsandwich committed:
Use "basic" format for ISO8601 date to match basic-formatted time.

The expanded time format uses colon characters which are not valid in
all filesystems. Rather than using the extended format for date and
basic for time, use the basic format for both.

nuclearsandwich committed:
Add timestamp to build results archive.

nuclearsandwich commented:
@Ronoman while I'm waiting for your review would you prefer me merge this PR after that review or would you prefer to be the one who merges PRs on this project for the time being?

The ROS 2 core team's convention is that team members merge their own PRs (with a "you break it you bought it" motivation) but I'm happy to do otherwise if you prefer

Ronoman commented on docs/build-results-archive.md:
```diff
@@ -0,0 +1,75 @@
+### processed-sarif-cmd
+
+This file is optional but recommended.
+If it is present, the processed.sarif file _must_ also be present.
```
Is this left behind from a previous iteration? My understanding is we decided for multiple `.sarif` files (corroborated below in processed/**/*.sarif), but this might be for something else that I'm forgetting about.

nuclearsandwich replied:
corrected in e7eded4

Ronoman commented on process_sarif/build_results_archive.py:
```diff
@@ -0,0 +1,156 @@
+
+        sarif_files = get_sarif_in_build(verbose=False)
+
+        remove_duplicate_results(sarif_files)
```
Does this work? My intent with this function is that it doesn't modify its inputs and instead returns `List[SarifFile]` that have duplicated removed, but its possible that [the function](https://github.com/space-ros/process_sarif/blob/main/process_sarif/sarif_helpers.py#L88) does in fact modify the argument in-place.

nuclearsandwich replied:
Good catch! This is left over from one of my earlier drafts fixing up the remove_duplicates function which did modify in place. ebdca25 saves the processed files and writes _them_ to the processed subdir instead of the original SARIF files.

Ronoman reviewed (changes requested):
Two very small changes to address before merging, but this looks great.

Ronoman commented:
Since these are your changes, I'll let you merge the PR. In the future I don't have a preference, but the ROS2 core team convention seems reasonable to me.

nuclearsandwich committed:
remove_build_results does not modify SarifFiles in place.

nuclearsandwich committed:
Update doc mentioning processed sarif which is no longer consolidated to a single file.
